### ISSUE TITLE
[Linux] Please make media file links case insensitive

### ISSUE THREAD

**Issue by uhrheber** [UltraStar-Deluxe/USDX#425]
### Actual behaviour

Some song creators (accidentally or carelessly) use different cases in the links to the media files than the real file names.
Example: MySong.txt:
#MP3:The best Band in the World.mp3

Real file name: The Best Band In The World.mp3

Windows doesn't care about this, and finds the files anyway (and that's why the creator never notices the error), but Linux does, and therefore doesn't find the file.

### Expected behaviour

The Linux version should either ignore the cases of the file names by default, to match the Windows behaviour, or at least have an option to do so. Correcting the song files manually is a lot of work for big collections.

### Steps to reproduce

Change one letter of a songs media file from upper to lower case or vice versa, start USDX, see the Error about a missing media file in the log.

### Details

- USDX version: Every Linux version up to today
- Operating System + version: generic Linux.

**Comment by bohning** [comment]
As a quick fix, you could use [UltraStar Manager](https://github.com/UltraStar-Deluxe/UltraStar-Manager) (use 'assign missing tasks automatically' task).

**Comment by s09bQ5** [comment]
What should USDX do on Linux if there are several files that only differ in case?

**Comment by uhrheber** [comment]
@bohning: I didn't know that it can do that, it needs about half an hour to open my song collection, therefore I never used it. This function indeed assigned most of the missing files automatically. Many thanks.

**Comment by ePirat** [comment]
> Try to find the one that matches the case exactly, if there is none, write a warning to the log, then convert everything to lower case and use the first find.

This can't work for case-sensitive file systems. Imagine the file is called "Foo.mp3" and the USDX song file has "FOO.mp3" instead. Converting that to all lowercase would give "foo.mp3" which still does not match "Foo.mp3". USDX would need to try all variations of lower/upper-case for each character in the file name which is a huge performance impact.

**Comment by abrock** [comment]
> USDX would need to try all variations of lower/upper-case for each character in the file name which is a huge performance impact.

No, it would only need to convert all file names in USDX song files to lowercase and store for each file found in the song directory both the original path and the path converted to lowercase. The performance impact is most likely insignificant compared to for example the time needed for filesystem traversal.

**Comment by basisbit** [comment]
@abrock feel free to submit a pull request, having fun with freepascal ;-)
Because of how song file scanning is implemented in usdx, this would either require quite a bit of refactoring/rewriting or it would require a full folder contents lookup for each valid looking us .txt file, increasing loading time up to a third on HDD systems with inefficient caching / big ultrastar song collections

**Comment by ePirat** [comment]
> No, it would only need to convert all file names in USDX song files to lowercase

I don't understand how this is supposed to work, sorry. As @basisbit said, feel free to submit a pull request that implements it.

**Comment by basisbit** [comment]
@ePirat as far as I understood, @abrock suggested that (enhancement suggestion: if the file is not found) usdx looks up all file names in that folder, then tries to find the matching file and then use the correct file name internally. On second thought, this might not be too bad after all because that code could also be used to find files with non-ansi-characters in filename on windows or other systems

**Comment by ePirat** [comment]
Ah indeed that should work if the file list is not too big and given how USDX songs are usually structured it should be relatively small in most cases. Clever idea!

**Comment by jose1711** [comment]
related and a possible workaround: https://www.phoronix.com/scan.php?page=news_item&px=EXT4-Case-Insensitive-Sent

**Comment by dgruss** [comment]
this is not an issue to address in usdx but a matter of configuration preferences for linux file systems

**Comment by barbeque-squared** [comment]
the one suggested by abrock will still produce undefined results if there are multiple files matching. Example:
- txt wants `Foo.mp3`
- actual files are `foo.mp3` and `FOO.mp3`

which one will it pick? or is this a Windows-only feature (because you can't copy both foo and FOO there anyway and you'll get interesting errors if copying from a linux user) but even then, we're trying to get rid of platform differences, not add new ones.

I feel this is mostly an issue if you copy a badly-maintained collection from a Windows user, and then subsequently don't contribute _back_ the correct txt's so people stop perpetuating broken txt's (that, or the Windows user doesn't integrate it back into their collection because "it works on my machine")

If anything, I'd rather find a way to make them behave case-sensitive on Windows as well. WSL apparently already has options for this on a per-directory level, so I wouldn't be surprised if applications in general start converging to case sensitive behaviour in the future.

**Comment by dgruss** [comment]
if you have for instance a shared folder and sync songs between linux and windows, and on Linux you have `foo.mp3` and `FOO.mp3` in the same folder, the syncing software will complain or crash. similar with git repos. On Windows you run into problems with that and we cannot solve the problem that Windows and Linux behave differently wrt. that. Or as you describe, a collection that is somehow else transferred across machines.
And that's not the only issue. Maybe the Windows user called the config file CONFIG.INI and now it does not use the same config when copied to the Linux machine although the user intended that.
This is a Linux vs Windows issue outside of the scope of usdx.

if Windows converges to case sensitive this issue also solves itself...

**Comment by ePirat** [comment]
Not only is this a Windows vs Linux issue but it also depends on the file system. On macOS you can have case-sensitive or case-insensitive file systems for example so this is not really something we can do much about.

**Comment by dgruss** [comment]
this is nothing the app can solve in any meaningful way --> should be closed
